hail/python/test/hail/expr/test_expr.py

```diff
@@ -274,25 +274,56 @@ def test_aggregators(self):
 
-    def test_aggregator_maps(self):
+    def test_new_aggregator_maps(self):
         t = hl.utils.range_table(10)
-        tests = [(agg.filter(lambda x: x > 8, agg._map(lambda x: x + 1, t.idx)), [9, 10]),
-                 (agg._map(lambda x: x + 1, agg.filter(lambda x: x > 8, t.idx)), [10]),
-                 (agg._flatmap(lambda x: hl.cond(x > 8, [x, x + 1], hl.empty_array(hl.tint32)), agg._map(lambda x: x + 1, t.idx)), [9, 10, 10, 11]),
-                 (agg._map(lambda x: x + 1, agg._flatmap(lambda x: hl.cond(x > 8, [x, x + 1], hl.empty_array(hl.tint32)), t.idx)), [10, 11]),
-                 (agg.filter(lambda x: x > 8, agg._flatmap(lambda x: [x, x + 1], t.idx)), [9, 9, 10]),
-                 (agg._flatmap(lambda x: [x, x + 1], agg.filter(lambda x: x > 8, t.idx)), [9, 10])]
-
 
+        tests = [(agg.filter(t.idx > 7,
+                             agg.collect(t.idx + 1).append(0)),
+                  [9, 10, 0]),
+                 (agg.explode(lambda elt: agg.collect(elt + 1).append(0),
+                              hl.cond(t.idx > 7, [t.idx, t.idx + 1], hl.empty_array(hl.tint32))),
+                  [9, 10, 10, 11, 0]),
+                 (agg.explode(lambda elt: agg.explode(lambda elt2: agg.collect(elt2 + 1).append(0),
+                                                      [elt, elt + 1]),
+                              hl.cond(t.idx > 7, [t.idx, t.idx + 1], hl.empty_array(hl.tint32))),
+                  [9, 10, 10, 11, 10, 11, 11, 12, 0]),
+                 (agg.explode(lambda elt: agg.filter(elt > 8,
+                                                     agg.collect(elt + 1).append(0)),
+                              hl.cond(t.idx > 7, [t.idx, t.idx + 1], hl.empty_array(hl.tint32))),
+                  [10, 10, 11, 0]),
+                 (agg.filter(t.idx > 7,
+                             agg.explode(lambda elt: agg.collect(elt + 1).append(0),
+                                         [t.idx, t.idx + 1])),
+                  [9, 10, 10, 11, 0]),
+                 (agg.group_by(t.idx % 2,
+                               hl.array(agg.collect_as_set(t.idx + 1)).append(0)),
+                  {0: [1, 3, 5, 7, 9, 0], 1: [2, 4, 6, 8, 10, 0]}),
+                 (agg.group_by(t.idx % 3,
+                               agg.filter(t.idx > 7,
+                                          hl.array(agg.collect_as_set(t.idx + 1)).append(0))),
+                  {0: [10, 0], 2: [9, 0]}),
+                 # {0: [10, 0], 1: [0], 2: [9, 0]}),
+                 #FIXME: I'm not super clear on how this distinction would work
+                 # currently, because the KeyedAggregator stuff appears to pass
+                 # through the filters and explodes. I think a way around this
+                 # for the time being would be to add a count aggregator just
+                 # for the purposes of getting the right keys, but we can also
+                 # leave this as is until we fix up the IR next week.
+                 (agg.filter(t.idx > 7,
```

Comment:
group_by/explode an explode/group_by would be good.  I think if we have all the pair-wise interactions, we should be good.

Also tests with multiple aggregators (including a multiple aggregators wrapped in the same filter that has randomness), e.g. `hl.agg.filter(hl.rand_bool(0.5), hl.tuple([hl.agg.sum(t.idx), hl.agg.sum(t.idx)]))`.